tests/tracer/test_logging.py

```diff
@@ -0,0 +1,356 @@
+        assert re.search(LOG_PATTERN, content) is not None
+
+    code = """
+import logging
+
+ddtrace_logger = logging.getLogger('ddtrace')
+assert ddtrace_logger.level == logging.WARN
+assert len(ddtrace_logger.handlers) == 1
+assert isinstance(ddtrace_logger.handlers[0], logging.handlers.RotatingFileHandler)
+assert ddtrace_logger.handlers[0].maxBytes == 200000
+assert ddtrace_logger.handlers[0].backupCount == 1
+
+ddtrace_logger.warning('warning log')
+"""
+
+    out, err, status, pid = ddtrace_run_python_code_in_subprocess(code, env=env)
+    assert status == 0, err
+    assert err == b""
+    assert out == b""
+    with open(log_file) as file:
+        content = file.read()
+        assert len(content) > 0
+        assert "warning log" in content
+        assert re.search(LOG_PATTERN, content) is not None
+
+
+def test_debug_logs_go_to_stderr(run_python_code_in_subprocess, ddtrace_run_python_code_in_subprocess):
+    """
+    When setting up the default logger in debug mode without a log path,
+        it automatically logs to stderr.
+    """
+    env = os.environ.copy()
+    env["DD_TRACE_DEBUG"] = "true"
+    code = """
+import logging
+import ddtrace
+
+ddtrace_logger = logging.getLogger('ddtrace')
+assert ddtrace_logger.level == logging.DEBUG
+assert len(ddtrace_logger.handlers) == 1
+assert isinstance(ddtrace_logger.handlers[0], logging.StreamHandler)
+"""
+
+    out, err, status, pid = run_python_code_in_subprocess(code, env=env)
+    assert status == 0, err
+    assert re.search(LOG_PATTERN, str(err)) is not None
+    assert b"debug logs" in err
+    assert out == b""
+
+    code = """
+import logging
+
+ddtrace_logger = logging.getLogger('ddtrace')
+assert ddtrace_logger.level == logging.DEBUG
+assert len(ddtrace_logger.handlers) == 1
+assert isinstance(ddtrace_logger.handlers[0], logging.StreamHandler)
+"""
+
+    out, err, status, pid = ddtrace_run_python_code_in_subprocess(code, env=env)
+    assert status == 0, err
+    assert re.search(LOG_PATTERN, str(err)) is not None
+    assert b"debug logs" in err
+    assert out == b""
+
+
+def test_debug_logs_can_go_to_file(run_python_code_in_subprocess, ddtrace_run_python_code_in_subprocess, tmpdir):
+    """
+    When setting up the default ddtrace logger in debug mode,
+        it automatically logs to a file at the expected backup count.
+    """
+    env = os.environ.copy()
+    log_file = tmpdir + "/testlog.log"
+    env["DD_TRACE_LOG_FILE"] = log_file
+    env["DD_TRACE_DEBUG"] = "true"
+    env["DD_TRACE_FILE_SIZE_BYTES"] = "10"
+    code = """
+import logging
+import ddtrace
+
+ddtrace_logger = logging.getLogger('ddtrace')
+assert ddtrace_logger.level == logging.DEBUG
+assert len(ddtrace_logger.handlers) == 1
+assert isinstance(ddtrace_logger.handlers[0], logging.handlers.RotatingFileHandler)
+assert ddtrace_logger.handlers[0].maxBytes == 10
+assert ddtrace_logger.handlers[0].backupCount == 1
+"""
+
+    out, err, status, pid = run_python_code_in_subprocess(code, env=env)
```

Comment:
I am just thinking out loud, since we have the duplicate assertions, I wonder if we should create helper functions or just iterate over the subprocess runners:

```python
for runner in [run_python_code_in_subprocess, ddtrace_run_python_code_in_subprocess]:
    out, err, status, pid = runner(code, env=env)
    assert status == 0, err
    assert err == b""
    assert out == b""
    
    # more assertions
```

I wouldn't say this is super clean, but could DRY this up a little bit... ?

Reply:
The `code` block used in each variation is slightly different. With the new changes to `_logger.py`, it turns out there is a slight difference in logs due to `commands/ddtrace_run` for ddtrace_run_python_code_in_subprocess.

See the latest commit for `assert "program executable" in str(err)` in some of the `ddtrace-run` tests.

If I went with the above approach, this would look more like this, I think it could be DRY still, but easier to read:

```
for runner in [run_python_code_in_subprocess, ddtrace_run_python_code_in_subprocess]:
   
    out, err, status, pid = runner(code, env=env)
    
    # common assertions
    assert status == 0, err
    assert err == b""
    assert out == b""
    
    # if assertion is ddtrace_run_python_code_in_subprocess:
    # ddtrace_run_python_code_in_subprocess specific assertions
    
    
```

Would it be fine to have runner specific asserts if it was refactored with the iterator? If so, I can make that update instead!